firedrake/paraview_reordering.py

```diff
@@ -0,0 +1,449 @@
+                   [2, 1, 3], [1, 0, 3]]
+    FaceMinCoord = [1, 3, 0, 2]
+    bindex = [0, 0, 0, 0]
+    # Can this condition ever fire?
+    while (index >= 2 * (order * order + 1) and index != 0 and order > 3):
+        index -= 2 * (order * order + 1)
+        idxmax -= 3
+        idxmin += 1
+        order -= 4
+    if (index < 4):
+        # we are on a vertex
+        for i in range(4):
+            bindex[i] = (idxmax if VertexMaxCoords[index] == i else idxmin)
+        return bindex
+    elif ((index - 4) < 6 * (order - 1)):
+        # we are on an edge
+        edgeId = (index - 4) // (order - 1)
+        vertexId = (index - 4) % (order - 1)
+        for coord in range(4):
+            temp1 = LinearVertices[EdgeVertices[edgeId][0]][coord]
+            temp2 = (idxmax - idxmin - 1 - vertexId)
+            temp3 = LinearVertices[EdgeVertices[edgeId][1]][coord]
+            temp4 = (1 + vertexId)
+            bindex[coord] = idxmin + (temp1 * temp2 + temp3 * temp4)
+
+        return bindex
+    else:
+        # we are on a face
+        faceId = (index - 4 - 6 * (order - 1)) // ((order - 2) * (order - 1) // 2)
+        vertexId = (index - 4 - 6 * (order - 1)) % ((order - 2) * (order - 1) // 2)
+        pbindex = [0, 0, 0]
+        if (order != 3):
+            pbindex = triangle_barycentric_index(vertexId, order - 3)
+        for i in range(3):
+            bindex[FaceBCoords[faceId][i]] = (idxmin + 1 + pbindex[i])
+        bindex[FaceMinCoord[faceId]] = idxmin
+        return bindex
+
+
+def bar_to_cart_3d(bar):
+    v0 = np.array([0, 0, 0])
+    v1 = np.array([1, 0, 0])
+    v2 = np.array([0, 1, 0])
+    v3 = np.array([0, 0, 1])
+    mat = np.array([v1, v2, v3, v0])
+    return np.dot(bar, mat)
+
+
+def vtk_tet_local_to_cart(order):
+    r"""Produces a list of nodes for VTK's lagrange tet basis.
+    :arg order: the order of the tet
+    :return a list of arrays of floats
+    """
+    count = int((order + 1) * (order + 2) * (order + 3) // 6)
+    bars = [np.array(tet_barycentric_index(i, order))/order for i in range(count)]
+    carts = [bar_to_cart_3d(b) for b in bars]
+    return carts
+
+
+def qsynatx(test, t, f):
+    if test:
+        return t
+    else:
+        return f
+
+
+def vtk_hex_point_index_from_ijk(i, j, k, order=None):
+    r"""
+    See vtkLagrangeHexahedron::PointIndexFromIJK
+    """
+    ibdy = (i == 0 or i == order[0])
+    jbdy = (j == 0 or j == order[1])
+    kbdy = (k == 0 or k == order[2])
+
+    itrue = i != 0
+    jtrue = j != 0
+    ktrue = k != 0
+
+    nbdy = int(jbdy) + int(kbdy) + int(ibdy)
+
+    if nbdy == 3:
+        # return vertex
+        # interprets:  (i ? (j ? 2 : 1) : (j ? 3 : 0)) + (k ? 4 : 0);
+        ret = 4 if ktrue else 0
+        if itrue:
+            ret += 2 if jtrue else 1
+        else:
+            ret += 3 if jtrue else 0
+        return ret
+    offset = 8
+    if nbdy == 2:  # edge
+        if not ibdy:  # on the i axis
+            temp0 = i - 1
```

Comment:
I don't think any of these temporaries are doing anything for us. Just add onto the offset.